extensions/amp-smartlinks/0.1/amp-smartlinks.js

```diff
@@ -0,0 +1,195 @@
+
+    /** @private {?Object} */
+    this.linkmateOptions_ = null;
+
+    /** @private {?function({JsonObject})} */
+    this.fetchLinkmateConfig_ = null;
+
+    /** @private {?../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = null;
+
+    /** @private {?./linkmate.Linkmate} */
+    this.linkmate_ = null;
+
+    /** @private {?string} */
+    this.referrer_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.xhr_ = Services.xhrFor(this.win);
+    this.ampDoc_ = this.getAmpDoc();
+
+    this.viewer_ = Services.viewerForDoc(this.ampDoc_);
+
+    this.linkmateOptions_ = getConfigOptions(this.element);
+    this.fetchLinkmateConfig_ = this.getLinkmateOptions_;
+
+    this.linkRewriterService_ = new LinkRewriterManager(this.ampDoc_);
+
+    return this.ampDoc_.whenBodyAvailable()
+        .then(() => this.viewer_.getReferrerUrl())
+        .then(referrer => {
+          this.referrer_ = referrer;
+          this.runSmartlinks_();
+        });
+  }
+
+  /**
+   * Wait for the config promise to resolve and then proceed to functionality
+   * @private
+   */
+  runSmartlinks_() {
+    this.fetchLinkmateConfig_().then(config => {
+      this.linkmateOptions_.linkmateExpected = config['linkmate_enabled'];
+      this.linkmateOptions_.publisherID = config['publisher_id'];
+
+      this.postPageImpression_();
+      this.linkmate_ = new Linkmate(
+          this.ampDoc_,
+          this.xhr_,
+          this.linkmateOptions_
+      );
+    });
+
+    this.smartLinkRewriter_ = this.initLinkRewriter_();
+
+    // We have a valid linkmate class, the config specified linkmate to run
+    // and our API is expecting linkmate to run
+    if (this.linkmate_ && this.linkmateOptions_.linkmateEnabled &&
+      this.linkmateOptions_.linkmateExpected) {
+      this.smartLinkRewriter_.getAnchorReplacementList();
+    }
+  }
+
+  /**
+   * API call to retrieve the Narrativ config for this extension.
+   * @return {?Promise}
+   * @private
+   */
+  getLinkmateOptions_() {
+    const fetchUrl = ENDPOINTS.NRTV_CONFIG_ENDPOINT.replace(
+        '.nrtv_slug.', this.linkmateOptions_.nrtvSlug
+    );
+
+    return this.xhr_.fetchJson(fetchUrl, {
+      method: 'GET',
+      ampCors: false,
+    })
+        .then(res => res.json())
+        .then(res => {
+          return getData(res)[0]['amp_config'];
+        });
+  }
+
+  /**
+   * API call to indicate a page load event happened
+   * @private
+   */
+  postPageImpression_() {
+    const payload = this.buildPageImpressionPayload_();
+
+    this.xhr_.fetchJson(ENDPOINTS.PAGE_IMPRESSION_ENDPOINT, {
+      method: 'POST',
+      ampCors: false,
+      headers: dict({'Content-Type': 'application/json'}),
+      body: payload,
+    });
+  }
+
+  /**
+   * Initialize and register a Narrativ LinkRewriter instance
+   * @return {!./link-rewriter/link-rewriter.LinkRewriter}
+   * @private
+   */
+  initLinkRewriter_() {
+    const options = {linkSelector: this.linkmateOptions_.linkSelector};
```

Comment:
@PhilWinchester So I asked @erwinmombay offline. We don't think that closure compilier is smart enough to understand that the options is the same to the `LinkerRewriter` options params without a `typedef`. This won't be an issue right now because all fieldName is reserved today, but it will be an issue when closure compiler start to obfuscated the fieldName. 

We can either switch to using a JsonObject. But given only `linkSelector` is supported, can we instead pass a string to the constructor? If you think that makes sense I'm fine with a following PR to address the comment.

Reply:
@zhouyx 
Do you mean doing something like this? 
```js
    const options = /** @type {JsonObject} */
      ({'linkSelector': this.linkmateOptions_.linkSelector});
```
I also assume that this part of the extension will need to be changed to if/when the affiliate link changes are added to the AMP project and the `link-rewriter` service is deprecated. I'm not sure if that would happen sooner than the closure compiler changes. Thanks for your help!

Reply:
nvm the `link-rewriter` code isn't expecting a JsonObject. I think I will follow up later with these changes.

Reply:
Even with the affiliate link changes, we are going to heavily rely on the existing link-rewriter service. As said I'd imaging this class to be kept and reused.